### ISSUE TITLE
[Feature Request] Fall through pattern matching

### ISSUE THREAD

**Issue by TeoTN** [jfairbank/redux-saga-router#10]
Hi! I've noticed that `ruta3` library which `redux-saga-router` is based on supports matching multiple patterns to one location. It can be useful when we want to have some general saga to be applied alongside detailed ones. For instance:
```
const routes = {
  '/profile/:id/': getProfileData,
  '/profile/:id/friends/': getProfileFriends,
  '/profile/:id/projects/': getProfileProjects,
}
```
It is easy to imagine, that on every subpage of `/profile/:id` we'd like to have some basic data fetched when visiting for the first time (e.g. username, photo, etc.) and also some detailed data.

I'd suggest adding such feature as an optional setting - and I am issuing a pull request for that. 
BTW My pull request also includes updated `README.md` with detailed information on what patterns are available.

**Comment by jfairbank** [comment]
Hi, @TeoTN!

This looks like a potentially useful feature! Thanks for the issue and pull requests.

I'll review this issue and the PRs more thoroughly soon and get back to you. I may have some questions on this feature in particular that we'll need to discuss.

**Comment by jfairbank** [comment]
Thanks for the related PR @TeoTN. I wanted to keep the discussion around the feature in this issue.

The only downside I'm seeing is that we run the risk of rerunning a saga unnecessarily. I'm fuzzy on what the best experience needs to be here.

For example, assume you have the routes `/foo` and `/foo/:id`.

If you hit the `/foo` route first, then it runs obviously. Then, if you click a link to visit `/foo/1`, you'll run its saga but also force the `/foo` saga to run again. That may not be what you want, especially if the `/foo` saga makes an unnecessary HTTP request again. I'm wondering if this is the behavior we want, or if it makes sense to track currently matched routes to avoid rerunning them when the route changes. Therefore, if we go from `/foo` to `/foo/1` it would know to only run the `/foo/:id` saga and not the `/foo` saga again.

Obviously, if you were on the `/foo/1` route, visited a `/bar` route, and then came back to the `/foo/1` route, then you would want the `/foo` and `/foo/:id` sagas to both run.

Does that make sense? What are your thoughts?

**Comment by TeoTN** [comment]
I think I see your point here but my assumption was that having set explicitly the fall-through behaviour we want to manually check whether we have to do such request at all. That is, higher-level saga (i.e. the one matched earlier) would check whether the state is incomplete. 

In particular, there may exist a case where the programmer may be interested in having the request performed anyway, although I have difficulties with justifying it properly. What comes to my mind is such case: the `/profile/` path is responsible for loading player up-to-date score, whereas `/profile/settings` would load his current name & surname (assuming that it wasn't already fetched) to fill the form. And if the score is subject to frequent changes on multiple devices and if we don't benefit (for any reason) from websockets then such additional request could be desirable. 

Another story is that it'd be a kind of 'conditional fall through' which may be slightly confusing - the option would cause the matching to fall through all prefix-matches but if and only if the prefix has changed. (And what if we have multiple nested routes?)

**Comment by jfairbank** [comment]
> Another story is that it'd be a kind of 'conditional fall through' which may be slightly confusing

I think at this time, it would be better to avoid the extra complexity as you mention. I may want to explore that behavior in the future, but let's keep it simple for now. We'll need to make sure devs are aware of that behavior, so they can write their sagas appropriately.

What are your thoughts on renaming the option, though? I'm not a huge fan of `shouldFallThrough`. What do you think of `matchAll` instead?

**Comment by TeoTN** [comment]
I like the `matchAll`, it's concise and clear :)

**Comment by jfairbank** [comment]
Released in v2.1.0. Thanks again, @TeoTN!
